### PR TITLE
Literal search by default

### PR DESCRIPTION
**Pull request by attfarhan**
This PR implements [RFC 31: Literal search by default](https://docs.google.com/document/d/11avb0ixXHGmHrpUM1-1ijSxogWiYMXXH1SCM8ozvX-s/edit#).

The changes made in this PR are as follows:

**GraphQL API**

This PR updates the `search` GraphQL API field to accept two new, optional parameters: `version` and `patternType`. 

`version` defaults to V1 if not passed, so existing scripts, src-cli, and other clients that aren't upgraded will continue to work. 

`patternType` is an optional parameter passed, which will override `version` when determining which pattern type to use. There is a new enum called `SearchPatternType` which specs the valid values, which are currently `regexp` and `literal`.

**Go backend**

The search GraphQL API resolver was updated to accept the `version` and `patternType` parameters. The resolver now checks whether a query should be interpreted as literal or regular expression. It does this via checking versions (V1 = regexp, V2 = literal). Then, it checks whether the `patternType` parameter has a value, and uses this to override the configuration it determined via the `version` parameter. Finally, it calls the new `HandlePatternType()` function, which checks for the `patternType:` field in the query, and if so, will override both the `version` and `patternType` parameters.

`HandlePatternType()` will then convert the query from regex to literal if necessary. 

**Webapp**

The webapp now has a new query parameter `patternType=`, which determines the pattern type to be used for a query on `search` routes. This is the source of truth for what will be passed as the `patternType` parameter to the `search` GraphQL endpoint, for search requests coming from the webapp. All searches on the webapp will have a `patternType=` query parameter.

The search input now has a regexp toggle, which, when clicked, will toggle the `patternType=` query parameter in the URL.

There is also a new setting, `search.defaultPatternType`, which can configure the default pattern type for an instance or user.

Because the `patternType` query parameter is now required for all search routes, the `buildSearchURLQuery` function has been updated to take an additional `patternType` parameter. This forces all components to specify the patternType before navigating to the search results page.

The `buildSearchURLQuery` function will check if the `patternType:` field exists in the raw query, and if so, will use that value in the `patternType=` URL query parameter, overriding the `patternType` argument passed into the function. This means that queries with the `patternType:` field in them will have the filter removed from the query, and the value will be reflected in the URL query parameter.

For URLs that do not have a `patternType=` query parameter, the `SearchResults` page component will detect that, and redirect to a URL appending `patternType=regexp`. This ensures that old links continue to work. This also updates the regexp toggle to accurately show the active state.

There is also a new pop-up toast `LiteralSearchToast` that appears for a user visiting the instance for the first time. The pop-up tells the user that there is a new regexp toggle, and that searches are now non-regexp by default, and links to the docs page. Once dismissed, it will not appear again for the user.

**Browser extension**

The browser extension's omnibar search is updated to pass `literal` as the patternType parameter to to `buildSearchURLQuery`. Omnibar search users can override this by including `patternType:regexp` in their query.

**Saved searches**

A migration was added to append `patternType:regexp` to all existing saved searches on startup. This is to avoid saved searches breaking due to any ambiguity about the patternType, which could arise if the instance/user defaults are changed.

The createSavedSearch and updateSavedSearch endpoints will throw an error, displayed in the saved search form, if a user tries to create a saved search without a `patternType` parameter.

**Testing**:
- I added e2e tests for the regexp toggle.
- There are unit tests for the Go backend.
- Existing search tests were updated, with either "patternType:regexp" or "V1" passed to maintain the old testing behavior.

I would encourage reviewers to pull down this branch and stress test any of these changes since it is a wide surface area.

### CODE REVIEW

File: web/src/marketing/LiteralSearchToast.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        return (
+            <div className="e2e-literal-search-toast">
+                <Toast
+                    icon={<RegexIcon />}
+                    title="New regular expression toggle!"
+                    subtitle="Search queries are no longer interpreted as regular expressions by default. Click the dot-star icon in the search bar to toggle regular expression search. Learn more about this update in the docs."
+                    onDismiss={this.onDismiss}
+                    cta={
+                        <Link to="/user/search/queries" className="btn btn-primary mr-2">
```
Comment (chrismwendt) [inline web/src/marketing/LiteralSearchToast.tsx]: This redirected me to http://localhost:3080/users/test/search/queries on my local machine, which 404ed.

It looks like this is done elsewhere like so:

https://github.com/sourcegraph/sourcegraph/blob/a02833213c6fdc6faeb0a95e87ab9c7d96ce1818/web/src/search/input/QueryBuilder.tsx#L133

https://github.com/sourcegraph/sourcegraph/blob/a02833213c6fdc6faeb0a95e87ab9c7d96ce1818/web/src/search/input/QueryBuilder.tsx#L372

File: web/src/search/input/RegexpToggle.scss [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    top: 0.375rem;
+    right: 0.5rem;
+    cursor: pointer;
+
+    &--active {
+        border: 1px solid $color-border-active;
+        padding-bottom: 0.25rem;
+        border-radius: 2px;
+    }
+}
```
Comment (chrismwendt) [inline web/src/search/input/RegexpToggle.scss]: Immediately after turning regex OFF, a faint halo remains, which looks very similar to regex ON:

image

Regex ON (personal preference: I'd like the whole square to be highlighted a light shade of blue):

image

Regex OFF:

image

File: web/src/search/helpers.tsx [hunk 3]
```diff
@@ -13,10 +13,12 @@ export function submitSearch(
     history: H.History,
     query: string,
     source: 'home' | 'nav' | 'repo' | 'tree' | 'filter' | 'type',
+    patternType: GQL.SearchPatternType,
     activation?: ActivationProps['activation']
 ): void {
     // Go to search results page
-    const path = '/search?' + buildSearchURLQuery(query)
+    console.log('@@@pattern type', patternType)
```
Comment (chrismwendt) [inline web/src/search/helpers.tsx]: Unintentional `console.log`?

File: .vscode/settings.json [hunk 4]
```diff
@@ -42,4 +42,5 @@
       "autoFix": true,
     },
   ],
+  "go.inferGopath": false,
```
Comment (lguychard) [inline .vscode/settings.json]: Why is this change necessary?

File: shared/src/util/url.ts [hunk 5]
```diff
@@ -534,11 +535,31 @@ export function withWorkspaceRootInputRevision(
 /**
  * Builds a URL query for the given query (without leading `?`).
  */
-export function buildSearchURLQuery(query: string): string {
+export function buildSearchURLQuery(query: string, patternType: SearchPatternType): string {
     const searchParams = new URLSearchParams()
-    searchParams.set('q', query)
+    const patternTypeInQuery = parsePatternTypeFromQuery(query)
+    if (patternTypeInQuery) {
```
Comment (lguychard) [inline shared/src/util/url.ts]: It would be useful to document in the docstring that if a pattern type is present in the query, it, will override the `patternType` argument. `buildSearchURLQuery()` unit tests should also test this.

File: web/src/search/input/RegexpToggle.tsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import React from 'react'
+import * as H from 'history'
+import RegexIcon from 'mdi-react/RegexIcon'
+import { submitSearch } from '../helpers'
+import { SearchPatternType } from '../../../../shared/src/graphql/schema'
+
+interface RegexpToggleProps {
+    togglePatternType: (patternType: SearchPatternType) => void
```
Comment (lguychard) [inline web/src/search/input/RegexpToggle.tsx]: I would consider the following:

```typescript
interface RegexpToggleProps {
    toggled: boolean,
    togglePatternType: () => void
    ...
}

// In parent component, pass in toggled={patternType === 'regex'} as props
```

This would simplify the checks on `patternType` in this component (which would then probably be best rewritten as a function component).

File: web/src/search/input/ScopePage.tsx [hunk 7]
```diff
@@ -35,6 +35,8 @@ const ScopeNotFound: React.FunctionComponent = () => (
 
 interface ScopePageProps extends RouteComponentProps<{ id: GQL.ID }>, SettingsCascadeProps {
     authenticatedUser: GQL.IUser | null
+    patternType: GQL.SearchPatternType
```
Comment (lguychard) [inline web/src/search/input/ScopePage.tsx]: Consider declaring `PatternTypeProps` once, and and referencing it across declarations.

File: web/src/search/input/RegexpToggle.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    public render(): JSX.Element | null {
+        return (
+            <button
+                onClick={this.toggle}
+                className="btn btn-icon icon-inline regexp-toggle e2e-regexp-toggle"
+                type="button"
+            >
+                <span
+                    className={`${
+                        this.props.patternType === 'regexp' ? 'regexp-toggle--active e2e-regexp-toggle--active' : ''
```
Comment (lguychard) [inline web/src/search/input/RegexpToggle.tsx]: You don't need to wrap this in a template string.

File: web/src/savedSearches/SavedSearchCreateForm.tsx [hunk 9]
```diff
@@ -63,7 +63,11 @@ export class SavedSearchCreateForm extends React.Component<Props, State> {
         const q = new URLSearchParams(this.props.location.search)
         let defaultValue: Partial<SavedQueryFields> = {}
         const query = q.get('query')
-        if (query) {
+        const patternType = q.get('patternType')
+
+        if (query && patternType) {
+            defaultValue = { query: query + ` patternType:${patternType}` }
```
Comment (lguychard) [inline web/src/savedSearches/SavedSearchCreateForm.tsx]: Should this account for the query containing a `patternType`?

File: web/src/search/results/SearchResults.tsx [hunk 10]
```diff
@@ -124,7 +149,11 @@ export class SearchResults extends React.Component<SearchResultsProps, SearchRes
                                                         : results.cloning.length > 0,
                                                 },
                                             },
-                                        }),
+                                        })
+                                        if (patternType && patternType !== this.props.patternType) {
+                                            this.props.togglePatternType(patternType)
```
Comment (lguychard) [inline web/src/search/results/SearchResults.tsx]: I would expect a function named `toggleSomething()` to not take arguments, and to simply flip the value of `something`. If it makes sense for the function to take an argument (because, for example, `something` ids an union of > 2 elements), it should be named `setSomething(newValue)`.

File: CHANGELOG.md [hunk 11]
```diff
@@ -23,6 +26,10 @@ All notable changes to Sourcegraph are documented in this file.
 - Indexed search won't spam the logs on startup if the frontend API is not yet available. [zoekt#30](https://github.com/sourcegraph/zoekt/pull/30), [#5866](https://github.com/sourcegraph/sourcegraph/pull/5866)
 - Sourcegraph now creates a secondary database within the configured PostgreSQL instance. This database has a name of the form `{PGDATABASE}_lsif`, where `{PGDATABASE}` is the name of the primary database. This is done automatically, but please ensure the configured PostgreSQL user has permission to perform `CREATE DATABASE` if using an external database before upgrading.
 - Search query fields are now case insensitive. For example `repoHasFile:` will now be recognized, not just `repohasfile:`. [#5168](https://github.com/sourcegraph/sourcegraph/issues/5168)
+- Search queries are now interpreted literally by default, rather than as regular expressions.
```
Comment (christinaforney) [inline CHANGELOG.md]: Add the issue link to this one.

```suggestion
- Search queries are now interpreted literally by default, rather than as regular expressions.  [#5899](https://github.com/sourcegraph/sourcegraph/pull/5899)
```

File: web/src/marketing/LiteralSearchToast.tsx [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+    public render(): JSX.Element | null {
+        if (!this.state.visible) {
+            return null
+        }
+
+        return (
+            <div className="e2e-literal-search-toast">
+                <Toast
+                    icon={<RegexIcon />}
```
Comment (christinaforney) [inline web/src/marketing/LiteralSearchToast.tsx]: Make sure the icon is top aligned with the title text, and that it looks visually centered in the column. You can also reduce the space between the icon and the content. Increase the padding between the text and the learn more button (I think 1rem looks good). Something like this:
image

File: web/src/marketing/LiteralSearchToast.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+                    title="New regular expression toggle!"
+                    subtitle="Search queries are no longer interpreted as regular expressions by default. Click the dot-star icon in the search bar to toggle regular expression search. Learn more about this update in the docs."
```
Comment (christinaforney) [inline web/src/marketing/LiteralSearchToast.tsx]:
```suggestion
                    subtitle="Search queries are no longer interpreted as regexp by default. Click the dot-star icon in the search bar to toggle regexp search. Learn more about this update in the docs."
```

File: web/src/savedSearches/SavedSearchCreateForm.tsx (hunk 9 above)
Comment (attfarhan) [reply web/src/savedSearches/SavedSearchCreateForm.tsx]: It already does. Default values get set if the user comes to the saved search form from the search results page. On the search results page, the patternType filter would've already been removed from the query and put in the patternType query parameter.

Comment (lguychard) [reply web/src/savedSearches/SavedSearchCreateForm.tsx]: Got it, thanks!

File: web/src/search/input/RegexpToggle.scss (hunk 2 above)
Comment (attfarhan) [reply web/src/search/input/RegexpToggle.scss]: Thanks for the feedback! Updated to have a background color when active: 

image
image

cc @christinaforney

File: web/src/search/input/RegexpToggle.tsx (hunk 6 above)
Comment (attfarhan) [reply web/src/search/input/RegexpToggle.tsx]: Used this pattern but did not end up writing the component as a FunctionalComponent as `submitSearch` requires passing `history` as an argument, which is not available in the top-level SourcegraphWebApp component where the `patternType` state is held.

File: web/src/search/input/RegexpToggle.scss (hunk 2 above)
Comment (christinaforney) [reply web/src/search/input/RegexpToggle.scss]: Looks great! Can you check to be sure the contrast ratio is ok? The light mode looks ok, but I can't tell on the dark mode: https://webaim.org/resources/contrastchecker/

Comment (attfarhan) [reply web/src/search/input/RegexpToggle.scss]: Updated to this, which is more obvious (>2:1 ratio):
image

Comment (christinaforney) [reply web/src/search/input/RegexpToggle.scss]: Looks better, thanks!

File: web/src/marketing/LiteralSearchToast.tsx [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+        return (
+            <div className="e2e-literal-search-toast">
+                <Toast
+                    icon={<RegexIcon />}
+                    title="New regular expression toggle!"
+                    subtitle="Search queries are no longer interpreted as regexp by default. Click the dot-star icon in the search bar to toggle regexp search. Learn more about this update in the docs."
+                    onDismiss={this.onDismiss}
+                    cta={
+                        <Link to={`${docsURLPrefix}/user/search/queries`} className="btn btn-primary mr-2">
```
Comment (christinaforney) [inline web/src/marketing/LiteralSearchToast.tsx]: This needs a label so it is clear to a screen reader what they are learning more about.

File: web/src/marketing/LiteralSearchToast.tsx [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+    public render(): JSX.Element | null {
+        const docsURLPrefix = this.props.isSourcegraphDotCom ? 'https://docs.sourcegraph.com' : '/help'
+        if (!this.state.visible) {
+            return null
+        }
+
+        return (
+            <div className="e2e-literal-search-toast">
+                <Toast
```
Comment (christinaforney) [inline web/src/marketing/LiteralSearchToast.tsx]: The toast dismissible 'x' button needs to have an aria label added saying "close" or something.

File: web/src/search/input/RegexpToggle.tsx [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+                .subscribe(event => {
+                    event.preventDefault()
+                    this.toggle()
+                })
+        )
+    }
+
+    public render(): JSX.Element | null {
+        return (
+            <div
```
Comment (christinaforney) [inline web/src/search/input/RegexpToggle.tsx]: Needs aria label as well so the screen reader can describe. Right now the screen reader says "checkbox, checked" but we don't know what the checkbox is for.

image

**Comment by slimsag** [comment]
you have a conflict in migrations/bindata.go btw

File: CHANGELOG.md [hunk 17]
```diff
@@ -15,6 +15,10 @@ All notable changes to Sourcegraph are documented in this file.
 
 - Our external service syncing model will stream in new repositories to Sourcegraph. Previously we could only add a repository to our database and clone it once we had synced all information from all external services (to detect deletions and renames). Now adding a repository to an external service configuration should be reflected much sooner, even on large instances. [#5145](https://github.com/sourcegraph/sourcegraph/issues/5145)
 - There is now an easy way for site admins to view and export settings and configuration when reporting a bug. The page for doing so is at /site-admin/report-bug, linked to from the site admin side panel under "Report a bug".
+- Search queries can now be interpreted literally.
+  - There is now a dot-star icon in the search input bar to toggle the pattern type of a query between regexp and literal.
```
Comment (slimsag) [inline CHANGELOG.md]:
```suggestion
- There is now a dot-star icon in the search input bar to toggle the pattern type of a query between regexp and literal.
```

File: CHANGELOG.md [hunk 18]
```diff
@@ -15,6 +15,10 @@ All notable changes to Sourcegraph are documented in this file.
...
+  - There is a new `search.defaultPatternType` setting to configure the default pattern type, regexp or literal, for searches.
```
Comment (slimsag) [inline CHANGELOG.md]:
```suggestion
- There is a new `search.defaultPatternType` setting to configure the default pattern type, regexp or literal, for searches.
```

File: CHANGELOG.md [hunk 19]
```diff
@@ -15,6 +15,10 @@ All notable changes to Sourcegraph are documented in this file.
...
+  - There is a new `patternType:` search token which overrides the `search.defaultPatternType` setting, and the active state of the dot-star icon in determining the pattern type of the query.
```
Comment (slimsag) [inline CHANGELOG.md]:
```suggestion
- There is a new `patternType:` search token which overrides the `search.defaultPatternType` setting, and the active state of the dot-star icon in determining the pattern type of the query.
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -15,6 +15,10 @@ All notable changes to Sourcegraph are documented in this file.
 
 - Our external service syncing model will stream in new repositories to Sourcegraph. Previously we could only add a repository to our database and clone it once we had synced all information from all external services (to detect deletions and renames). Now adding a repository to an external service configuration should be reflected much sooner, even on large instances. [#5145](https://github.com/sourcegraph/sourcegraph/issues/5145)
 - There is now an easy way for site admins to view and export settings and configuration when reporting a bug. The page for doing so is at /site-admin/report-bug, linked to from the site admin side panel under "Report a bug".
+- Search queries can now be interpreted literally.
```
Comment (slimsag) [inline CHANGELOG.md]:
```suggestion
- The default search mode is now literal search instead of regular expression search. This can be toggled in the UI, and configured in the settings (see below).
```

File: cmd/frontend/graphqlbackend/saved_searches.go [hunk 21]
```diff
@@ -271,3 +280,10 @@ func (r *schemaResolver) DeleteSavedSearch(ctx context.Context, args *struct {
 	}
 	return &EmptyResponse{}, nil
 }
+
+func queryHasPatternType(query string) bool {
+	patternTypeRegexp := regexp.MustCompile(`(?i)\bpatternType:(literal|regexp)\b`)
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/saved_searches.go]: Define this as a variable outside the function and then you won't pay the cost of compiling the regexp every time.

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 22]
```diff
@@ -1005,6 +1005,11 @@ type Query {
     clientConfiguration: ClientConfigurationDetails!
     # Runs a search.
     search(
+        # The version of the search syntax being used.
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
        # The version of the search syntax being used.
        # All new clients should use the latest version.
```

File: cmd/frontend/graphqlbackend/search_results_test.go [hunk 23]
```diff
@@ -145,7 +154,78 @@ func TestSearchResults(t *testing.T) {
...
+		}
+		defer func() { mockSearchRepositories = nil }()
+
+		calledSearchSymbols := false
+		mockSearchSymbols = func(ctx context.Context, args *search.Args, limit int) (res []*fileMatchResolver, common *searchResultsCommon, err error) {
+			calledSearchSymbols = true
+			if want := `"foo\\d \"bar*\""`; args.Pattern.Pattern != want {
+				t.Errorf("got %q, want %q", args.Pattern.Pattern, want)
+			}
+			// TODO return mock results here and assert that they are output as results
```
Comment (slimsag) [inline cmd/frontend/graphqlbackend/search_results_test.go]: bump

File: cmd/frontend/internal/pkg/search/query/pattern_type.go [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+	"strings"
+)
+
+var fieldRx = regexp.MustCompile(`^-?[a-zA-Z]+:`)
+
+// HandlePatternType returns a modified version of the input query where it has
+// been either quoted because it has patternType:literal, not quoted because it
+// has patternType:regex, or been possibly quoted in the default case according
+// to the defaultToRegexp parameter.
+func HandlePatternType(input string, defaultToRegexp bool) string {
```
Comment (slimsag) [inline cmd/frontend/internal/pkg/search/query/pattern_type.go]: You can follow up on this later, but: You shouldn't be using regex here for tokenization, we have an existing query parser and I am pretty sure you could (or should be able to, at least) reuse it or this purpose.

File: web/src/e2e/e2e.test.ts [hunk 25]
```diff
@@ -1163,11 +1168,55 @@ describe('e2e test suite', () => {
                 expect(numberOfResults).toBeGreaterThan(0)
             })
         })
+        // TODO should we wait for an element to prevent flakiness
```
Comment (slimsag) [inline web/src/e2e/e2e.test.ts]: bump

**Comment by slimsag** [comment]
image

In this message, please:

1. Replace `dot-star` with the actual icon if possible.
2. Spell out "regular expressions" instead of writing "regexp".

**Comment by slimsag** [comment]
This is SO cool! This is gonna be such a massive improvement for users!

**Comment by slimsag** [comment]
uh oh, I visited an old URL I had bookmarked, http://localhost:3080/search?q=err.*or and it redirected me to http://localhost:3080/search?q=err.*or&patternType=literal which gives no results. This might be a downfall not covered in the RFC, but it did call out as a goal:

> Existing Sourcegraph search links do not break, and sharing links between users works regardless of the instance- or user-level search type default (literal or regexp).

File: CHANGELOG.md (hunk 20 above)
Comment (attfarhan) [reply CHANGELOG.md]: Having a top-level bullet and 3 sub-bullets was a suggestion from @christinaforney to group these items together (see https://github.com/sourcegraph/sourcegraph/pull/5899#discussion_r333355192), I quite like this format, but realize it's the only sub-bullet list. Do you have a strong objection to this @slimsag?

**Comment by attfarhan** [comment]
Stephen and I chatted, and we think a good solution for old links is to add `&patternType=regexp` to URLs that don't have a `patternType=`. parameter. There is no other case (except perhaps URL hacking, which is tough to do because you need to escape characters in your query, etc.) where URLs don't have a patternType appended to it (omnibar searches default to `patternType=literal`).

File: web/src/nav/GlobalNavbar.tsx [hunk 26]
```diff
@@ -37,6 +38,8 @@ interface Props
      * homepage.
      */
     lowProfile: boolean
+    patternType: GQL.SearchPatternType
```
Comment (lguychard) [inline web/src/nav/GlobalNavbar.tsx]: Use `PatternTypeProps` here.

File: cmd/frontend/internal/pkg/search/query/pattern_type.go (hunk 24 above)
Comment (keegancsmith) [reply cmd/frontend/internal/pkg/search/query/pattern_type.go]: @attfarhan I just found this PR after reviewing https://github.com/sourcegraph/sourcegraph/pull/6256/files

I agree with this feedback in general. Transforming the input into a string which is then parsed by our original parser seems like a way for us to run into plenty more bugs down the line. I'll continue the discussion in the new PR, but just pointing out that @slimsag's point was very important.

Meta note: This PR is massive! In future try and break down the problem a bit so we can incrementally review and catch these higher level improvements.
